panic in rules api

### What happened?

Seems to be an error in the merge algorithm


```
panic: expected struct or pointer to a struct got ptr                                                                                                                                                   
                                                                                                                                                                                                       
goroutine 1663 [running]:                                                                                                                                                                               
github.com/kumahq/kuma/pkg/core/xds.mustUnwrapStruct({0x27adb20?, 0xc000bde468?, 0x8?})                                                                                                                 
    /home/circleci/project/pkg/core/xds/merge.go:115 +0xb4                                                                                                                                              
github.com/kumahq/kuma/pkg/core/xds.appendSlices({0x27adb20?, 0xc000bdedb0?, 0x0?}, {0x27adb20?, 0xc000bde468?, 0x27adb20?})                                                                            
    /home/circleci/project/pkg/core/xds/merge.go:87 +0x9b                                                                                                                                               
github.com/kumahq/kuma/pkg/core/xds.appendSlices({0x23bbc00?, 0xc000bdedb0?, 0xb0?}, {0x2311a60?, 0xc000e4d250?, 0x30?})                                                                                
    /home/circleci/project/pkg/core/xds/merge.go:103 +0x31f                                                                                                                                             
github.com/kumahq/kuma/pkg/core/xds.MergeConfs({0xc002038160, 0x2, 0x0?})                                                                                                                               
    /home/circleci/project/pkg/core/xds/merge.go:49 +0x4af                                                                                                                                              
github.com/kumahq/kuma/pkg/core/xds.BuildRules({0xc00206d940, 0x2, 0xc000ed4e70?})                                                                                                                      
    /home/circleci/project/pkg/core/xds/rules.go:162 +0x570                                                                                                                                             
github.com/kumahq/kuma/pkg/plugins/policies/matchers.fromRules(0xc00155b328?)                                                                                                                           
    /home/circleci/project/pkg/plugins/policies/matchers/match.go:98 +0xff                                                                                                                              
github.com/kumahq/kuma/pkg/plugins/policies/matchers.MatchedPolicies({0x282755a, 0xb}, 0xc001563710, {0xc001941fb0?, 0xc000f66180?})                                                                    │
    /home/circleci/project/pkg/plugins/policies/matchers/match.go:55 +0x2fd                                                                                                                             
github.com/kumahq/kuma/pkg/plugins/policies/meshtimeout/plugin/v1alpha1.plugin.MatchedPolicies(...)                                                                                                     
    /home/circleci/project/pkg/plugins/policies/meshtimeout/plugin/v1alpha1/plugin.go:34                                                                                                                
github.com/kumahq/kuma/pkg/xds/sync.(*DataplaneProxyBuilder).matchPolicies(0xc0003b3020?, {{0xc001662e40, 0x2c}, 0xc0016ed338, {0xc001941fb0, 0xc000f66180}, 0xc001738db0, 0xc001739410, 0xc001739560,  │
    /home/circleci/project/pkg/xds/sync/dataplane_proxy_builder.go:165 +0x791                                                                                                                           
github.com/kumahq/kuma/pkg/xds/sync.(*DataplaneProxyBuilder).Build(0xc0003b3020, {0x32096e0, 0xc000a730c0}, {{0xc0007abd70?, 0xc0007d9fc0?}, {0xc0007abd78?, 0xc0007abd70?}}, {{0xc001662e40, 0x2c}, 0x │
    /home/circleci/project/pkg/xds/sync/dataplane_proxy_builder.go:45 +0x14b                                                                                                                            
github.com/kumahq/kuma/pkg/xds/sync.(*DataplaneWatchdog).syncDataplane(0xc001aecd20, {0x32096e0, 0xc000a730c0}, 0xc00252c320)                                                                           
    /home/circleci/project/pkg/xds/sync/dataplane_watchdog.go:117 +0x42e                                                                                                                                
github.com/kumahq/kuma/pkg/xds/sync.(*DataplaneWatchdog).Sync(0xc001aecd20, {0x32096e0, 0xc000a730c0})                                                                                                  
    /home/circleci/project/pkg/xds/sync/dataplane_watchdog.go:66 +0x171                                                                                                                                 
github.com/kumahq/kuma/pkg/xds/sync.(*dataplaneWatchdogFactory).New.func2()                                                                                                                             
    /home/circleci/project/pkg/xds/sync/dataplane_watchdog_factory.go:46 +0xc7                                                                                                                          
github.com/kumahq/kuma/pkg/util/watchdog.(*SimpleWatchdog).Start(0xc000766820, 0xc001434660)                                                                                                            
    /home/circleci/project/pkg/util/watchdog/watchdog.go:25 +0xe2                                                                                                                                       
created by github.com/kumahq/kuma/pkg/xds/server/callbacks.(*dataplaneSyncTracker).OnProxyConnected                                                                                                     
    /home/circleci/project/pkg/xds/server/callbacks/dataplane_sync_tracker.go:55 +0x40a
```

```bash
echo 'kind: MeshTimeout
apiVersion: kuma.io/v1alpha1
metadata:
  namespace: kuma-system
  name: from-all
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: Mesh
      default:
        connectionTimeout: 15s
        idleTimeout: 1h
        http:
          requestTimeout: 15s
          streamIdleTimeout: 1s
          maxStreamDuration: 0s
          maxConnectionDuration: 0s' | kc apply -f-
```

```bash
echo 'kind: MeshTimeout
apiVersion: kuma.io/v1alpha1
metadata:
  namespace: kuma-system
  name: fronted-to-backend
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          streamIdleTimeout: 2s
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      default:
        connectionTimeout: 15s
        idleTimeout: 1h
        http:
          requestTimeout: 15s
          streamIdleTimeout: 3s
          maxStreamDuration: 0s
          maxConnectionDuration: 0s' | kc apply -f-
```

--- a/pkg/core/xds/merge.go
+++ b/pkg/core/xds/merge.go
@@ -63,6 +63,9 @@ func newConf(t reflect.Type) (interface{}, error) {
 
 func clearAppendSlices(val reflect.Value) {
 	strVal := mustUnwrapStruct(val)
+	if strVal == reflect.ValueOf(nil) {
+		return
+	}
 	for i := 0; i < strVal.NumField(); i++ {
 		valField := strVal.Field(i)
 		fieldName := strVal.Type().Field(i).Name
@@ -85,6 +88,9 @@ func clearAppendSlices(val reflect.Value) {
 func appendSlices(dst reflect.Value, src reflect.Value) {
 	strDst := mustUnwrapStruct(dst)
 	strSrc := mustUnwrapStruct(src)
+	if strSrc == reflect.ValueOf(nil) || strDst == reflect.ValueOf(nil) {
+		return
+	}
 	for i := 0; i < strDst.NumField(); i++ {
 		dstField := strDst.Field(i)
 		srcField := strSrc.Field(i)
@@ -109,6 +115,9 @@ func appendSlices(dst reflect.Value, src reflect.Value) {
 func mustUnwrapStruct(val reflect.Value) reflect.Value {
 	resVal := val
 	if val.Kind() == reflect.Interface || val.Kind() == reflect.Pointer {
+		if val.IsNil() {
+			return reflect.ValueOf(nil)
+		}
 		resVal = val.Elem()
 	}
 	if resVal.Kind() != reflect.Struct {

--- a/pkg/core/xds/merge_test.go
+++ b/pkg/core/xds/merge_test.go
@@ -18,71 +18,185 @@ var _ = Describe("MergeConfs", func() {
 		Strings       []string
 		AppendStrings []string
 		Sub           subPolicy
-		SubPtr        *subPolicy
+		SubPtr        *subPolicy `json:"subPtr,omitempty"`
 	}
 
-	It("should replace slices by default but append slices that start with append", func() {
-		// given
-		p1 := policy{
-			FieldString:   "p1",
-			Strings:       []string{"p1"},
-			AppendStrings: []string{"p1"},
-			Sub: subPolicy{
-				Ints:       []int{1},
-				AppendInts: []int{1},
-			},
-			SubPtr: &subPolicy{
-				Ints:       []int{1},
-				AppendInts: []int{1},
-			},
-		}
+	type testCase struct {
+		policies []policy
+		expected policy
+	}
 
-		p2 := policy{
-			FieldString:   "p2",
-			Strings:       []string{"p2"},
-			AppendStrings: []string{"p2"},
-			Sub: subPolicy{
-				Ints:       []int{2},
-				AppendInts: []int{2},
-			},
-			SubPtr: &subPolicy{
-				Ints:       []int{2},
-				AppendInts: []int{2},
-			},
-		}
+	DescribeTable("merge confs",
+		func(given testCase) {
+			// given
+			var policies []interface{}
+			for _, policy := range given.policies {
+				policies = append(policies, policy)
+			}
 
-		p3 := policy{
-			Strings:       []string{"p3"},
-			AppendStrings: []string{"p3"},
-			Sub: subPolicy{
-				Ints:       []int{3},
-				AppendInts: []int{3},
-			},
-			SubPtr: &subPolicy{
-				Ints:       []int{3},
-				AppendInts: []int{3},
-			},
-		}
+			// when
+			merged, err := xds.MergeConfs(policies)
 
-		policies := []interface{}{p1, p2, p3}
-
-		// when
-		merged, err := xds.MergeConfs(policies)
-
-		// then
-		Expect(err).ToNot(HaveOccurred())
-		Expect(merged.(policy)).To(Equal(policy{
-			FieldString:   "p2",
-			Strings:       []string{"p3"},
-			AppendStrings: []string{"p1", "p2", "p3"},
-			Sub: subPolicy{
-				Ints:       []int{3},
-				AppendInts: []int{1, 2, 3},
+			// then
+			Expect(err).ToNot(HaveOccurred())
+			Expect(merged.(policy)).To(Equal(given.expected))
+		},
+		Entry("should replace slices by default but append slices that start with append", testCase{
+			policies: []policy{
+				{
+					FieldString:   "p1",
+					Strings:       []string{"p1"},
+					AppendStrings: []string{"p1"},
+					Sub: subPolicy{
+						Ints:       []int{1},
+						AppendInts: []int{1},
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{1},
+						AppendInts: []int{1},
+					},
+				},
+				{
+					FieldString:   "p2",
+					Strings:       []string{"p2"},
+					AppendStrings: []string{"p2"},
+					Sub: subPolicy{
+						Ints:       []int{2},
+						AppendInts: []int{2},
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{2},
+						AppendInts: []int{2},
+					},
+				},
+				{
+					Strings:       []string{"p3"},
+					AppendStrings: []string{"p3"},
+					Sub: subPolicy{
+						Ints:       []int{3},
+						AppendInts: []int{3},
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{3},
+						AppendInts: []int{3},
+					},
+				},
 			},
-			SubPtr: &subPolicy{
-				Ints:       []int{3},
-				AppendInts: []int{1, 2, 3},
+			expected: policy{
+				FieldString:   "p2",
+				Strings:       []string{"p3"},
+				AppendStrings: []string{"p1", "p2", "p3"},
+				Sub: subPolicy{
+					Ints:       []int{3},
+					AppendInts: []int{1, 2, 3},
+				},
+				SubPtr: &subPolicy{
+					Ints:       []int{3},
+					AppendInts: []int{1, 2, 3},
+				},
 			},
-		}))
-	})
+		}),
+		Entry("should handle nils on destination side", testCase{
+			policies: []policy{
+				{
+					FieldString:   "p1",
+					Strings:       []string{"p1"},
+					AppendStrings: []string{"p1"},
+					Sub: subPolicy{
+						Ints:       []int{1},
+						AppendInts: []int{1},
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{1},
+						AppendInts: []int{1},
+					},
+				},
+				{
+					FieldString:   "p2",
+					Strings:       nil,
+					AppendStrings: nil,
+					Sub: subPolicy{
+						Ints:       nil,
+						AppendInts: nil,
+					},
+					SubPtr: nil,
+				},
+			},
+			expected: policy{
+				FieldString:   "p2",
+				Strings:       nil,
+				AppendStrings: []string{"p1"},
+				Sub: subPolicy{
+					Ints:       nil,
+					AppendInts: []int{1},
+				},
+				SubPtr: &subPolicy{
+					Ints:       []int{1},
+					AppendInts: []int{1},
+				},
+			},
+		}),
+		Entry("should handle nils on source side", testCase{
+			policies: []policy{
+				{
+					FieldString:   "p1",
+					Strings:       nil,
+					AppendStrings: nil,
+					Sub: subPolicy{
+						Ints:       nil,
+						AppendInts: nil,
+					},
+					SubPtr: nil,
+				},
+				{
+					FieldString: "p2",
+					Sub: subPolicy{
+						AppendInts: []int{2},
+					},
+					SubPtr: &subPolicy{
+						AppendInts: []int{2},
+					},
+				},
+				{
+					FieldString:   "p3",
+					Strings:       nil,
+					AppendStrings: nil,
+					Sub: subPolicy{
+						Ints:       nil,
+						AppendInts: nil,
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{3},
+						AppendInts: nil,
+					},
+				},
+				{
+					FieldString:   "p4",
+					Strings:       []string{"4"},
+					AppendStrings: []string{"4"},
+					Sub: subPolicy{
+						Ints:       []int{4},
+						AppendInts: []int{4},
+					},
+					SubPtr: &subPolicy{
+						Ints:       []int{4},
+						AppendInts: []int{4},
+					},
+				},
+			},
+			expected: policy{
+				FieldString:   "p4",
+				Strings:       []string{"4"},
+				AppendStrings: []string{"4"},
+				Sub: subPolicy{
+					Ints:       []int{4},
+					AppendInts: []int{2, 4},
+				},
+				SubPtr: &subPolicy{
+					Ints:       []int{4},
+					AppendInts: []int{2, 4},
+				},
+			},
+		}),
+	)
 })

